primer/test/Tests/Action/Prog.hs

```diff
@@ -425,31 +425,7 @@ unit_create_typedef_bad_5 =
           , astTypeDefNameHints = []
           }
    in progActionTest defaultEmptyProg [AddTypeDef (tcn "T") td] $
-        expectError (@?= TypeDefError "InternalError \"Duplicate names in one tydef: between parameter-names and constructor-names\"")
-
--- Forbid clash between type name and parameter name
-unit_create_typedef_bad_6 :: Assertion
```

Comment:
Before I answer, in `data T (C :: *) (T :: * -> *) = ...`, the parameter `T` is a recursive reference to the typedef `T` itself, is that correct?

Reply:
I think no good can come of this, however clever it might seem to allow names within a particular typedef scope to clash. It might be slightly annoying to Haskellers to have to write `data F a = MkF a` rather than `data F a = F a`, but I don't think the cleverness is worth the potential for confusion for beginners, or even intermediate students. I can imagine teachers needing to spend an awful lot of time explaining the difference between `T` as type name vs `T` as type parameter vs `T` as value constructor, all depending on which side of the `=` they're on, and/or whether they appear in head position or as a parameter. That doesn't seem like time spent constructively, when we could just instead outlaw these clashes.

So I'm very much in favor of not allowing name conflicts within a given typedef scope.

Once we have better support in the UI for displaying the type of a value constructor when it appears in an action input, I think it might be OK to treat value constructor names as being prefixed by their type and therefore allowing the same value constructor name in different types [*], but I assume that is not covered by this particular commit.

[*] In other words, at the moment if we have `data A = Foo | Bar` and `data B = Foo | Baz`, when the student selects the "Use value constructor" action, she is presented with `[Foo, Bar, Foo, Baz]` and there is no way to tell which `Foo` is which, so this would be bad. But if instead we presented her with `[Foo : A, Bar : A, Foo : B, Baz : B]` then I think we could make a case for allowing `Foo` in `A` and `B`.

Reply:
Ok, I'll change approach here to forbid such clashes again.

Reply:
> Ok, I'll change approach here to forbid such clashes again.

Done -- dropped that commit (not needed to pass testsuite, cf comment above) and added a new commit on the end to actually do the checks.

Reply:
> I assume that these checks were removed to make the property test pass easier, but expect that we could instead engineer the test to treat these like name capture and ignore these specific errors. (@georgefst could you confirm?)

Yes, that is the case. At the time I chose this approach, I wasn't sure how to ignore the errors. I am less sure than @dhess about the pros of outlawing such pseudo-clashes, but I'll write my thoughts about that up in a separate thread.